services/brig/test/unit/Test/Brig/Calling.hs

```diff
@@ -0,0 +1,206 @@
+{-# LANGUAGE RecordWildCards #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2020 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Test.Brig.Calling where
+
+import Brig.Calling
+import Brig.Options
+import Brig.PolyLog
+import Control.Retry
+import Data.List.NonEmpty (NonEmpty (..))
+import Imports
+import Network.DNS
+import Polysemy
+import qualified System.Logger as Log
+import Test.Tasty
+import Test.Tasty.HUnit
+import qualified UnliftIO.Async as Async
+import Wire.Network.DNS.Effect
+import Wire.Network.DNS.SRV
+
+data FakeDNSEnv = FakeDNSEnv
+  { fakeLookupFn :: Domain -> SrvResponse,
+    fakeLookupCalls :: IORef [Domain]
+  }
+
+newFakeDNSEnv :: (Domain -> SrvResponse) -> IO FakeDNSEnv
+newFakeDNSEnv lookupFn = do
+  FakeDNSEnv lookupFn <$> newIORef []
+
+runFakeDNSLookup :: Member (Embed IO) r => FakeDNSEnv -> Sem (DNSLookup ': r) a -> Sem r a
+runFakeDNSLookup FakeDNSEnv {..} = interpret $ \case
+  LookupSRV domain -> do
+    modifyIORef' fakeLookupCalls (++ [domain])
+    pure $ fakeLookupFn domain
+
+newtype LogRecorder = LogRecorder {recordedLogs :: IORef [(Log.Level, LByteString)]}
+
+newLogRecorder :: IO LogRecorder
+newLogRecorder = LogRecorder <$> newIORef []
+
+recordLogs :: Member (Embed IO) r => LogRecorder -> Sem (PolyLog ': r) a -> Sem r a
+recordLogs LogRecorder {..} = interpret $ \(PolyLog lvl msg) ->
+  modifyIORef' recordedLogs (++ [(lvl, Log.render (Log.renderDefault ", ") msg)])
+
+ignoreLogs :: Sem (PolyLog ': r) a -> Sem r a
```

Comment:
I don't like the name `PolyLog` (it's a nice name, but it doesn't point out what it refers to clearly enough for my taste).  But we can change it later.

I *do* like the fact that we're introducing polysemi!  :)